### PR TITLE
app/vmctl: add retry backoff policy

### PR DESCRIPTION
**Pull request by dmitryk-dk**
I moved the retry backoff policy to utils and used it in importer for different processors and native import process.
When we try to migrate data from OpenTSDB, InfluxDB, Remote read protocol, etc., the retries will start if we lose connection to the storage where we transfer data.

Remote read protocol migration process with retries
217252234-b192976d-b4c9-4d68-ae2c-bcd62db1c00d
217252264-85827f70-61f4-4160-b2dd-272ef0123a65

Related issue: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3600

Signed-off-by: dmitryk-dk [d.kozlov@victoriametrics.com](mailto:d.kozlov@victoriametrics.com)

### CODE REVIEW

File: app/vmctl/flags.go [hunk 1]
```diff
@@ -29,6 +29,32 @@ var (
...
 )
 
+const (
+	backoffRetries     = "backoff-retries"
+	backoffFactor      = "backoff-factor"
+	backoffMinDuration = "backoff-min-duration"
+)
+
+var (
+	retryFlags = []cli.Flag{
```
Comment (hagen1778) [inline app/vmctl/flags.go:39]: From the code below, looks like backoff flags are only used by VM importer and VM native. So, I think, these flags should belong to `vmFlags`.

File: app/vmctl/README.md [hunk 2]
```diff
@@ -1011,6 +1016,26 @@ The rate limit may be set in bytes-per-second via `--vm-rate-limit` flag.
...
+Retry backoff policy by default enabled on:
+- import process when migrate data from [Prometheus](#migrating-data-from-prometheus) to VictoriaMetrics using snapshot API
+- import process when migrate data from [Thanos](#migrating-data-from-thanos) to VictoriaMetrics
+- import process when migrate data from [Cortex](#migrating-data-from-cortex) to VictoriaMetrics
+- import process when migrate data from [Mimir](#migrating-data-from-mimir) to VictoriaMetrics
+- import process when migrate data from [InfluxDB](#migrating-data-from-influxdb-1x) to VictoriaMetrics
+- import process when migrate data from [OpenTSDB](#migrating-data-from-opentsdb) to VictoriaMetrics
+- export and import processes when migrate data between [VictoriaMetrics](#migrating-data-from-victoriametrics) single or cluster version.
+- import process when migrate data by [Prometheus remote read protocol](#migrating-data-by-remote-read-protocol) to VictoriaMetrics
+
```
Comment (hagen1778) [inline app/vmctl/README.md:1038]: I'm not sure we need all this complexity, to be fair. It is cool to have the backoff policy, and to have it configurable.
But for a couple of years that vmctl exists there was no a single request to make it adjustable. So I am thinking, do we need to add more flags, more documentation for this? Or we can just keep it inside with reasonable defaults? wdyt?

File: app/vmctl/vm/vm.go [hunk 3]
```diff
@@ -54,6 +55,13 @@ type Config struct {
 	RateLimit int64
 	// Whether to disable progress bar per VM worker
 	DisableProgressBar bool
+
+	// BackoffRetries defines how many retries we need to check if callback was successful.
+	BackoffRetries int
+	// BackoffFactor configure the length of delay after each failed attempt.
+	BackoffFactor float64
+	// BackoffMinDuration configure minimum (initial) repeat interval
+	BackoffMinDuration time.Duration
```
Comment (hagen1778) [inline app/vmctl/vm/vm.go:64]: Would it be easier to have a `backoff.Retry` object instead of these fields? In this way you won't need to describe these fields over and over for each new struct which needs retries.

File: app/vmctl/vm/vm.go [hunk 4]
```diff
@@ -219,7 +229,9 @@ func (im *Importer) startWorker(bar *pb.ProgressBar, batchSize, significantFigur
 			exitErr := &ImportError{
 				Batch: batch,
 			}
-			if err := im.Import(batch); err != nil {
+			retryableFunc := func() error { return im.Import(batch) }
+			_, err := im.retry.Do(context.Background(), retryableFunc)
```
Comment (hagen1778) [inline app/vmctl/vm/vm.go:233]: wdyt about supporting the real context here? So user can cancel it when stopping vmctl

File: app/vmctl/vm_native.go [hunk 5]
```diff
@@ -54,13 +59,75 @@ const (
...
+	requestLimitC := make(chan struct{}, p.requestsLimit)
+	filterCh := make(chan filter)
+	var wg sync.WaitGroup
+
+	defer func() {
+		close(filterCh)
+		wg.Wait()
+		close(requestLimitC)
+	}()
+
```
Comment (hagen1778) [inline app/vmctl/vm_native.go:102]: These changes seems irrelevant to the PR purpose

**Review by hagen1778 — CHANGES_REQUESTED** [review]

File: app/vmctl/flags.go [hunk 6]
```diff
@@ -29,6 +29,32 @@ var (
...
+		&cli.IntFlag{
+			Name:  backoffRetries,
+			Usage: "How many retries we need to check if callback was successful.",
```
[the remark this answers is not in the document]
Comment (dmitryk-dk) [reply app/vmctl/flags.go]: removed flags at all

File: app/vmctl/flags.go [hunk 7]
```diff
@@ -29,6 +29,32 @@ var (
...
+			Value: 5,
+		},
+		&cli.Float64Flag{
+			Name:  backoffFactor,
+			Usage: "Configure the length of delay after each failed attempt.",
```
[the remark this answers is not in the document]
Comment (dmitryk-dk) [reply app/vmctl/flags.go]: removed flags at all

File: app/vmctl/flags.go [hunk 8]
```diff
@@ -29,6 +29,32 @@ var (
...
+			Value: 1.7,
+		},
+		&cli.DurationFlag{
+			Name:  backoffMinDuration,
+			Usage: "Configure minimum (initial) repeat interval.",
```
[the remark this answers is not in the document]
Comment (dmitryk-dk) [reply app/vmctl/flags.go]: removed flags at all

File: app/vmctl/flags.go [hunk 9]
```diff
@@ -406,6 +433,18 @@ var (
...
+			Usage: "Number of workers concurrently performing import requests to VM",
+			Value: 2,
+		},
+		&cli.UintFlag{
+			Name: vmNativeRequestsLimit,
+			Usage: "Defines the number of concurrent export requests from the source database. It should be set to \n" +
+				" value of the -search.maxConcurrentRequests value specified on the vmselect\n" +
+				" if you use cluster version, or on vmsingle if you use single version.",
+			Value: 2,
+		},
```
[the remark this answers is not in the document]
Comment (dmitryk-dk) [reply app/vmctl/flags.go]: removed

File: app/vmctl/vm/vm.go (hunk 3 above)
Comment (dmitryk-dk) [reply app/vmctl/vm/vm.go]: used only object

File: app/vmctl/vm/vm.go (hunk 4 above)
Comment (dmitryk-dk) [reply app/vmctl/vm/vm.go]: used global context

File: app/vmctl/vm_native.go (hunk 5 above)
Comment (dmitryk-dk) [reply app/vmctl/vm_native.go]: will add this changes in the next improvement

File: app/vmctl/backoff/backoff.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	}
+}
+
+// Retry process retries until all attempts are completed
+func (b *Backoff) Retry(ctx context.Context, cb retryableFunc) (uint64, error) {
+	var attempt uint64
+	for i := 0; i < b.retries; i++ {
+		select {
+		case <-ctx.Done():
+			return attempt, nil
```
Comment (hagen1778) [inline app/vmctl/backoff/backoff.go:46]: this would work between retries, but will not cancel execution if we're in sleeping phase. Would be great to change this in future PRs. So far, `TODO` mark will be enough.

File: app/vmctl/backoff/backoff.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+		default:
+		}
+		err := cb()
+		if err == nil {
+			return attempt, nil
+		}
+		if errors.Is(err, ErrBadRequest) {
+			logger.Errorf("got error: %s on attempt: %d \n", err, attempt)
```
Comment (hagen1778) [inline app/vmctl/backoff/backoff.go:54]:
```suggestion
			logger.Errorf("unrecoverable error: %s", err)
```

File: app/vmctl/backoff/backoff.go [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+			return attempt, err // fail fast if not recoverable
+		}
+		attempt++
+		backoff := float64(b.minDuration) * math.Pow(b.factor, float64(i))
+		dur := time.Duration(backoff)
+		logger.Errorf("got error: %s on attempt: %d \n", err, attempt)
+		logger.Infof("next attempt will start after: %s \n", dur.Round(time.Millisecond))
```
Comment (hagen1778) [inline app/vmctl/backoff/backoff.go:61]:
```suggestion
		logger.Errorf("got error: %s on attempt: %d; will retry in %v", err, attempt, dur)
```

File: app/vmctl/backoff/backoff.go [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+		time.Sleep(time.Duration(backoff))
+	}
+	return attempt, fmt.Errorf("retry failed after %d retries", b.retries)
```
Comment (hagen1778) [inline app/vmctl/backoff/backoff.go:64]:
```suggestion
	return attempt, fmt.Errorf("execution failed after %d retry attempts", b.retries)
```

File: docs/CHANGELOG.md [hunk 14]
```diff
@@ -32,6 +32,7 @@ The following tip changes can be tested by building VictoriaMetrics components f
 * FEATURE: [MetricsQL](https://docs.victoriametrics.com/MetricsQL.html): add `range_trim_zscore(z, q)` function for dropping outliers located farther than `z*range_stddev(q)` from `range_avg(q)`. This should help removing outliers during query time at [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3759).
 * FEATURE: [vmui](https://docs.victoriametrics.com/#vmui): show `median` instead of `avg` in graph tooltip and line legend, since `median` is more tolerant against spikes. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3706).
 * FEATURE: allow setting zero value for `-search.latencyOffset` command-line flag. This may be needed in [some cases](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2061#issuecomment-1299109836). Previously the minimum supported value for `-search.latencyOffset` command-line flag was `1s`.
+* FEATURE: [vmctl](https://docs.victoriametrics.com/vmctl.html): add retry policy with backoff for different migration protocols. See [this feature request](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3600).
```
Comment (hagen1778) [inline docs/CHANGELOG.md:35]:
```suggestion
```

From what I understand, nothing is changed except refactoring. So it is not a feature.

**Review by hagen1778 — CHANGES_REQUESTED** [review]
Thanks! A few comments

File: app/vmctl/vm/vm.go [hunk 15]
```diff
@@ -264,30 +269,16 @@ func (im *Importer) startWorker(bar *pb.ProgressBar, batchSize, significantFigur
...
-		im.s.Lock()
-		im.s.retries++
-		im.s.Unlock()
-		backoff := float64(backoffMinDuration) * math.Pow(backoffFactor, float64(i))
-		time.Sleep(time.Duration(backoff))
+func (im *Importer) flush(ctx context.Context, b []*TimeSeries) error {
+	retryableFunc := func() error { return im.Import(b) }
+	attempts, err := im.backoff.Retry(ctx, retryableFunc)
+	if err != nil {
+		return fmt.Errorf("import failed with %d retries: %s", attempts, err)
```
Comment (hagen1778) [inline app/vmctl/vm/vm.go:276]: if we actually run this we'll get the following msg in logs:
```
ERROR import failed with 5 retries: execution failed after 5 retry attempts
```
So we provide the same info twice. I guess, we can simplify this a bit. But this can be done in a follow-up

**Review by hagen1778 — APPROVED** [review]
